pkg/image/api/types.go

```diff
@@ -60,6 +77,41 @@ type ImageLayer struct {
 	Size int64
 }
 
+// SignatureValidationResult holds a result of signature validation.
+type SignatureValidationResult string
+
+// ImageSignature holds a signature of an image.
+type ImageSignature struct {
+	// Describes a type of stored blob. Provided by caller.
+	Type string
+	// An opaque binary string which is an image's signature.
+	Content []byte
+	// A time of last server-side validation.
+	ValidatedOn unversioned.Time
+	// Result of validation performed at ValidatedOn.
+	ValidationResult SignatureValidationResult
+
+	// At most one of the following pointers can point to details. If the signature type isn't recognized,
+	// they will all be nil.
+
+	// Details for signature of type "atomic".
+	AtomicSignatureDetails *AtomicSignatureDetails
```

Comment:
If we have those, we should introduce fields that define them.  I don't see
value in having:

AtomicSpecificSignatureDetails
GPGSpecificSignatureDetails
FooSpceificSignatureDetails

because every UI has to then know about signature.  UIs don't *care* about
the types, they care about a set of fields.  A bunch of optional fields is
better than always forcing a set of optional structs.

This is about LCD.

On Wed, May 18, 2016 at 8:36 AM, David Eads <notifications@github.com>
wrote:

> In pkg/image/api/types.go
> <https://github.com/openshift/origin/pull/8371#discussion_r63694283>:
>
> > +// ImageSignature holds a signature of an image.
> > +type ImageSignature struct {
> > +	// Describes a type of stored blob. Provided by caller.
> > +	Type string
> > +	// An opaque binary string which is an image's signature.
> > +	Content []byte
> > +	// A time of last server-side validation.
> > +	ValidatedOn unversioned.Time
> > +	// Result of validation performed at ValidatedOn.
> > +	ValidationResult SignatureValidationResult
> > +
> > +	// At most one of the following pointers can point to details. If the signature type isn't recognized,
> > +	// they will all be nil.
> > +
> > +	// Details for signature of type "atomic".
> > +	AtomicSignatureDetails *AtomicSignatureDetails
>
> These are generic details core to the platform (image name). Unfilled
> fields can default to nothing. Having multiple structs for generic
> fields isn't useful.
>
> You're saying these examples in particular look generic to you, so you
> think nothing will ever be specific? Or just that these don't qualify?
>
> A signer could also sign something like a ruleset-version: 6 or human-readable-failure="security
> hole X", which has no generic representation, but in a first class
> integration could have significant utility.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/8371/files/a49f7a114201f3d8868dba2c837f84b85045e79c#r63694283>
>